File: cpp/19.顺时针打印矩阵.cpp
class Solution {
public:
    vector<int> printMatrix(vector<vector<int> > matrix) {
        vector<int> ret;
        if(matrix.empty())
            return ret;
        int up = 0, down = matrix.size(), left = 0, right = matrix[0].size();
        for(int i = 0, j = 0; up < down && left < right; left++, i++, j++){
            for(; left < right && up < down && j < right; j++)
                ret.push_back(matrix[i][j]);
            for(up++, i++, j--; left < right && up < down && i < down; i++)
                ret.push_back(matrix[i][j]);
            for(right--, i--, j--; left < right && up < down && j >= left; j--)
                ret.push_back(matrix[i][j]);
            for(down--, i--, j++; left < right && up < down && i >= up; i--)
                ret.push_back(matrix[i][j]);
        }
        return ret;
    }
};

File: java/19.顺时针打印矩阵.java
import java.util.ArrayList;
public class Solution {
    public ArrayList<Integer> printMatrix(int [][] matrix) {
       ArrayList<Integer> arr = new ArrayList<Integer>();
        if (matrix == null || matrix[0] == null)
           return arr;
        for (int i = 0, j = 0, up = 0, down = matrix.length - 1, left = 0, right = matrix[0].length - 1; left <= right && up <=down; i++, j++, left++) {
            for(; left <= right && up <= down && j <= right; arr.add(matrix[i][j++]));
            for(i++, j--, up++; left <= right && up <= down && i <= down; arr.add(matrix[i++][j]));
            for(j--, i--, right--; left <= right && up <= down && j >= left; arr.add(matrix[i][j--]));
            for(i--, j++, down--; left <= right && up <= down && i >= up; arr.add(matrix[i--][j]));
        }
        return arr;
    }
}
